src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,166 @@
+package com.minecolonies.coremod.colony;
+
+import com.minecolonies.coremod.util.Log;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+//TODO extend list/collection
+/**
+ * Data structure for storing colonies, optimized for performance.
+ *
+ * @author Colton
+ */
+public final class ColonyList<T extends IColony> implements Iterable<T>
+{
+    private static final int INITIAL_SIZE = 16;
+
+    private Object[] list = new Object[INITIAL_SIZE];
+
+    private List<Integer> nullIndices = new ArrayList<>();
+
+    private int topID = 0;
+
+    private int size = 0;
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public T get(int index) {
+        if (index < 1 || index > list.length) {
+            return null;
+        }
+
+        return (T) list[index];
+    }
+
+    @NotNull
+    public Colony create(World world, BlockPos position) {
+        Colony colony = new Colony(getNextColonyID(), world, position);
+        list[colony.getID()] = colony;
+        size++;
+        return colony;
+    }
+
+    public void add(T colony) {
+        T existingColony = get(colony.getID());
+        if (existingColony != null && existingColony != colony) {
+            throw new IllegalArgumentException(
+                    String.format("Already a colony registered to id=%d, colony=%s, not changing to colony=%s",
+                            colony.getID(),
+                            existingColony.getName(),
+                            colony.getName()));
+        }
+
+        list[colony.getID()] = colony;
+    }
+
+    public void remove(Colony colony) {
+        remove(colony.getID());
+    }
+
+    public void remove(int id) {
+        if (list[id] == null) {
+            Log.getLogger().warn("Tried to remove colony with id=%d, but it didn't exist.", id);
+        }
+
+        size--;
+        list[id] = null;
+
+        if (!nullIndices.contains(id)) {
+            nullIndices.add(id);
+        }
+    }
+
+    private int getNextColonyID() {
+        if (nullIndices.isEmpty()) {
+            topID++;
+            if (topID >= list.length) {
+                // Expand list
+                Object[] newList = new Object[list.length * 2];
+                System.arraycopy(list, 0, newList, 0, list.length);
+                list = newList;
+            }
+
+            return topID;
+        }
+
+        return nullIndices.remove(0);
+    }
+
+    public void clear() {
+        for (int i = 0; i < list.length; i++)
+        {
+            list[i] = null;
+        }
+
+        nullIndices.clear();
+
+        topID = 0;
+        size = 0;
+    }
+
+    public int size() {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Move this left curly brace to the beginning of next line of code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)